src/ts/proposal.ts

```diff
@@ -0,0 +1,181 @@
+import { MetaTransaction } from "@gnosis.pm/safe-contracts";
+import { HardhatEthersHelpers } from "@nomiclabs/hardhat-ethers/types";
+import { BigNumber, utils } from "ethers";
+
+import {
+  getDeterministicDeploymentTransaction,
+  getNonDeterministicDeploymentTransaction,
+  RealTokenDeployParams,
+  VirtualTokenDeployParams,
+  ContractName,
+} from "./deploy";
+import {
+  prepareDeterministicSafeWithOwners,
+  SafeDeploymentAddresses,
+  SafeOperation,
+} from "./lib/safe";
+import { metadata } from "./token";
+
+export interface SafeCreationSettings {
+  threshold: number;
+  owners: string[];
+  nonce?: string;
+}
+export interface RealTokenCreationSettings {
+  salt?: string;
+}
+export interface VirtualTokenCreationSettings {
+  merkleRoot: string;
+  usdcToken: string;
+  gnoToken: string;
+  gnoPrice: string;
+  wrappedNativeToken: string;
+  nativeTokenPrice: string;
+}
+export interface DeploymentProposalSettings {
+  cowDao: SafeCreationSettings;
+  teamController: SafeCreationSettings;
+  cowToken: RealTokenCreationSettings;
+  virtualCowToken: VirtualTokenCreationSettings;
+}
+
+export type JsonMetaTransaction = Record<
+  keyof Omit<MetaTransaction, "operation">,
+  string
+> & { operation: number };
+export interface FinalAddresses {
+  cowDao: string;
+  teamController: string;
+  investorFundsTarget: string;
+  cowToken: string;
+}
+export interface Proposal {
+  steps: JsonMetaTransaction[];
+  addresses: FinalAddresses;
+}
+
+export async function generateProposal(
+  settings: DeploymentProposalSettings,
+  safeDeploymentAddresses: SafeDeploymentAddresses,
+  ethers: HardhatEthersHelpers,
+): Promise<Proposal> {
+  const { address: cowDao, transaction: cowDaoCreationTransaction } =
+    await setupDeterministicSafe(
+      settings.cowDao,
+      safeDeploymentAddresses,
+      ethers,
+    );
+
+  const {
+    address: teamController,
+    transaction: teamControllerCreationTransaction,
+  } = await setupDeterministicSafe(
+    settings.teamController,
+    safeDeploymentAddresses,
+    ethers,
+  );
+
+  const {
+    address: investorFundsTarget,
+    transaction: investorFundsTargetCreationTransaction,
+  } = await setupDeterministicSafe(
+    { owners: [cowDao], threshold: 1 },
+    safeDeploymentAddresses,
+    ethers,
+  );
+
+  const realTokenDeployParams: RealTokenDeployParams = {
+    cowDao,
+    totalSupply: BigNumber.from(10).pow(3 * 4 + metadata.real.decimals),
```

Comment:
Should this be: 3 * 3 + metadata.real.decimals.

We want 1000M coins, right?

Reply:
Whoops!